sql/src/main/java/io/crate/executor/transport/TransportShardUpsertAction.java

```diff
@@ -321,22 +323,36 @@ public IndexRequest prepareUpdate(ShardUpsertRequest request, ShardUpsertRequest
     }
 
     private IndexRequest prepareInsert(ShardUpsertRequest request, ShardUpsertRequest.Item item) throws IOException {
-        BytesRef rawSource = null;
-        XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+        int rawSourceColumn = -1;
         for (int i = 0; i < item.insertValues().length; i++) {
             Reference ref = request.insertColumns()[i];
-            if (ref.info().ident().columnIdent().equals(DocSysColumns.RAW)) {
-                rawSource = (BytesRef)item.insertValues()[i];
+            if (ref.ident().columnIdent().equals(DocSysColumns.RAW)) {
+                rawSourceColumn = i;
                 break;
             }
-            builder.field(ref.ident().columnIdent().fqn(), item.insertValues()[i]);
         }
-        IndexRequest indexRequest = Requests.indexRequest(request.index()).type(request.type()).id(item.id()).routing(item.routing())
-                .create(!request.overwriteDuplicates()).operationThreaded(false);
-        if (rawSource != null) {
-            indexRequest.source(rawSource.bytes);
+
+        BytesReference source;
+        if (rawSourceColumn > -1) {
+            source = new BytesArray((BytesRef) item.insertValues()[rawSourceColumn]);
         } else {
-            indexRequest.source(builder.bytes());
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
```

Comment:
just to be safe, pls use XContentFactory.contentBuilder(Requests.INDEX_CONTENT_TYPE)